### PR TITLE
Finish Product Api

### PR DESCRIPTION
**Pull request by AyalaZisman**
product api and update the model

### CODE REVIEW

File: src/main/java/com/yellow/ordermanageryellow/Dao/AuditDataRepository.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.yellow.ordermanageryellow.Dao;
+
+import com.yellow.ordermanageryellow.model.AuditData;
+import org.springframework.data.mongodb.repository.MongoRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Dao/AuditDataRepository.java:7]: No need this repository

File: src/main/java/com/yellow/ordermanageryellow/Dao/ProductRepository.java [hunk 2]
```diff
@@ -3,6 +3,17 @@
...
+import org.springframework.stereotype.Repository;
 
+import java.util.List;
+
+@Repository
 public interface ProductRepository extends MongoRepository<Product, String> {
+boolean existsByName(String name);
+
+    @Query("{'name': { $regex: '^?1' }}")
+    List<Product> findByCompanyId_NameAndPrefix(String companyId, String prefix);
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Dao/ProductRepository.java:16]: you should find by comanyid not company name

File: src/main/java/com/yellow/ordermanageryellow/Dto/ProductDTO.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class ProductDTO {
+    private String id;
+    private String name;
+    private String desc;
+    private double price;
+    private String ProductCategoryId;
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Dto/ProductDTO.java:19]: lower latter

File: src/main/java/com/yellow/ordermanageryellow/Service/CompanyService.java [hunk 4]
```diff
@@ -8,15 +8,15 @@
...
-    public CompanyService(CompanyRepository CompanyRepository) {
-        this.CompanyRepository = CompanyRepository;
-    }
-    @Override
-    public void run(String... args) {
-        Company Company = new Company("12");
-        CompanyRepository.save(Company);
-    }
+public class CompanyService {
+//    private final CompanyRepository CompanyRepository;
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/CompanyService.java:12]: please removed commented lines

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductService.java [hunk 5]
```diff
@@ -1,23 +1,65 @@
...
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.CommandLineRunner;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
 import org.springframework.stereotype.Service;
 
+import java.time.LocalDate;
+import java.util.*;
+import java.util.stream.Collectors;
+
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/ProductService.java:21]: Plese format all your classes 
CTR + ALT + L

File: src/main/java/com/yellow/ordermanageryellow/controller/ProductController.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    public ResponseEntity getAllNamesProducts(@PathVariable("prefix") String prefix,@RequestHeader ("token")String token){
+        try{
+        return ResponseEntity.status(HttpStatus.OK).body(productService.getAllProductsNames(token,prefix)) ;
+    }
+    catch (Exception e ){
+            return  ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(e.getMessage()) ;
+        }
+    }
+    @GetMapping
+    public ResponseEntity getAllProduct(){
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/controller/ProductController.java:35]: you should retrieve products related to company

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductService.java [hunk 7]
```diff
@@ -1,23 +1,65 @@
...
+        Optional<Product> updatedProduct =this.productRepository.findById(product.getId());
+        if(updatedProduct.isEmpty())
+            throw new NoSuchElementException("product doesn't exist");
+       return this.productRepository.save(product);
+    }
+    public void deleteProduct(String id){
+        this.productRepository.deleteById(id);
     }
+    public List<ProductDTO> getAllProducts() {
+        List<Product> products = productRepository.findAll().stream().toList();
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/ProductService.java:58]: add company checks

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductService.java [hunk 8]
```diff
@@ -1,23 +1,65 @@
...
+public Product addProduct(Product product) throws ObjectAllReadyExists {
+    if (productRepository.existsByName(product.getName())==true)
+        throw new ObjectAllReadyExists("Object Is All Ready Exists");
+    product.setAuditData(new AuditData(LocalDate.now(), null));
+    Product newProduct= productRepository.insert(product);
+    return newProduct;
+
+}
+    public List<ProductNameDTO> getAllProductsNames(String token,String prefix){
+        List<Product> products = productRepository.findByCompanyId_NameAndPrefix(token,prefix).stream().toList();
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/ProductService.java:43]: token composed from usr role and company

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductService.java [hunk 9]
```diff
@@ -1,23 +1,65 @@
...
+        List<ProductNameDTO> productList = ProductMapper.INSTANCE.ProductToProductNameDTO(products);
+        return productList;
     }
-    @Override
-    public void run(String... args) {
-        Product Product = new Product("12");
-        ProductRepository.save(Product);
+
+    public Product editProduct(Product product){
+        Optional<Product> updatedProduct =this.productRepository.findById(product.getId());
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/ProductService.java:49]: please update auditdata.updatedate

**Review by moriyahazan — CHANGES_REQUESTED** [review]

File: src/main/java/com/yellow/ordermanageryellow/controller/ProductController.java (hunk 6 above)
Comment (AyalaZisman) [reply src/main/java/com/yellow/ordermanageryellow/controller/ProductController.java:35]: Ayala Shprey will add the token  for all of us. That's why I didn't do these tests

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductService.java (hunk 7 above)
Comment (AyalaZisman) [reply src/main/java/com/yellow/ordermanageryellow/Service/ProductService.java:58]: Ayala Sheprey will add the token horizontally for all of us. That's why I didn't do these tests

File: src/main/java/com/yellow/ordermanageryellow/Dao/ProductRepository.java [hunk 10]
```diff
@@ -10,10 +10,9 @@
 
 @Repository
 public interface ProductRepository extends MongoRepository<Product, String> {
-boolean existsByName(String name);
+    boolean existsByName(String name);
 
     @Query("{'name': { $regex: '^?1' }}")
-    List<Product> findByCompanyId_NameAndPrefix(String companyId, String prefix);
-
+    List<Product> findByCompanyIdAndNameAndPrefix(String companyId, String prefix);
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Dao/ProductRepository.java:16]: Are you sure this is working?? it should be  findByCompanyId_Id...

**Review by moriyahazan — APPROVED** [review]
Hi Ayala, 

Approved (I think there is problem with method  findByCompanyIdAnd.... please check)
please merge main to your branch

**Comment by AyalaZisman** [comment]
כן הפונק עובדת.
תודה רבה!

‫בתאריך יום א׳, 23 ביולי 2023 ב-2:58 מאת ‪Moriya‬‏ <‪
***@***.***‬‏>:‬

**Pull request merged**